### PR TITLE
[NavigationExperimental] Allow route to specify the animation direction in NavigationCardStack

### PR DESCRIPTION
**Pull request by cmcewen**
Explain the **motivation** for making this change. What existing problem does the pull request solve?

In using `NavigationCardStack`, there are some screens that I want to animate in from the side (viewing details on a piece of content), and some that I want to push from the bottom (like drafting a new piece of content). This provides a means of specifying the animation direction on a per route basis while still falling back to the overall default.

**Test plan (required)**

Create a NavigationCardStack and create routes with the `direction` property - see that the animation direction matches the specified direction.

### CODE REVIEW

File: Libraries/NavigationExperimental/NavigationTypeDefinition.js [hunk 1]
```diff
@@ -23,7 +23,8 @@ export type NavigationGestureDirection = 'horizontal' | 'vertical';
 
 export type NavigationRoute = {
   key: string,
-  title?: string
+  title?: string,
+  direction?: NavigationGestureDirection,
```
Comment (hedgerwang) [inline Libraries/NavigationExperimental/NavigationTypeDefinition.js]: Thanks for your contribution.

I'm concerned about that the route is becoming the convenient place that people could add anything to it.
I'd imagine that in few months later, the route will become something like this.

```
export type NavigationRoute = {
  key: string,
  title?: string,
  subTitle?: string,
  direction? (number|string),
  velocity? number,
  id? (number|string),
  blah: any,
  blahBlahBlah: any,
};
```

and so will people keep adding new things to `NavigationState`.

If this pattern is accepted, will this be scalable to support different gestures, animations, rendering needs?  

IMHO, we do allow people to add anything to the route, on the other hand, we don't want to limit the types of data that the route has or can carry.

Maybe we should explicit define a prop `meta` that can be anything thus people can add whatever type of data to it without restricting others. 

```
export type NavigationRoute = {
  key: string,
  meta?: any, 
};
```

Comment (cmcewen) [inline Libraries/NavigationExperimental/NavigationTypeDefinition.js]: @hedgerwang definitely a fair point - I'm already adding some props to pass to components in the route object and I could definitely see it growing in size to become unwieldy.

I haven't been following `NavigationExperimental` super closely so I'm not exactly sure the intent of the API design, but in this case I think it might make sense to add a few properties to the route if it is tightly coupled to a `NavigationCard`, like [`cardStyle`](https://github.com/facebook/react-native/blob/master/Libraries/CustomComponents/NavigationExperimental/NavigationCardStack.js#L65) and `direction`. Since the `NavigationCardStack` looks at these properties to render, I'm not sure that hiding it under another nested object would be clear.

Basically, if we're able to set some defaults for the whole card stack, it would also be nice to override those for specific cards - if there's a better place to do that then the route object then I'd definitely be interested in that, but from my quick reading of the code this seemed like the easiest place to make that happen.

**Comment by cmcewen** [comment]
@hedgerwang any more thoughts on this? My current `NavigationRoute` definition looks like this:
```
export type NavigationRoute = {
  key: string,
  title?: string,
  props?: Object,
  direction?: string,
};
```

**Comment by facebook-github-bot-1** [comment]
It's been a while since the last commit was reviewed and the labels show this pull request needs review. Based on the blame information for the files in this pull request we identified @hedgerwang as a potential reviewer. Could you take a look please or cc someone with more context?

**Comment by ericvicenti** [comment]
For the reason @hedgerwang pointed out, I don't think it is a best practice to put the direction on the navigation state.

Instead, what if we had something like this?

```javascript
<NavigationCardStack
  navigationState={...}
  renderScene={...}

  // Default:
  getCardComponent={scene => NavigationCard}

  // If you want to let your app define animations for each scene:
  getCardComponent={scene => MyRouter.getCardComponent(scene.route)}

  // And if you really want to use route 'direction':
  getCardComponent={scene => {
    switch (scene.route.direction) {
      case 'vertical': return VerticalNavigationCard;
      default: return NavigationCard;
    }
  }

...
